File: 0041/solution.ts
function firstMissingPositive(nums: number[]): number {
  const N = nums.length;
  let C = nums.length + 1;
  for (let i = 0; i < nums.length; i++) {
    if (nums[i] < 0 || nums[i] > N) {
      nums[i] = 0;
    }
  }
  for (let i = 0; i < nums.length; i++) {
    let loc = (nums[i] % C) - 1;
    if (loc >= 0) {
      nums[loc] = (nums[loc] % C) + C;
    }
  }
  for (let i = 0; i < nums.length; i++) {
    if (nums[i] < C) {
      return i + 1;
    }
  }
  return N + 1;
}

File: 0041/solution.py
class Solution:
    def firstMissingPositive(self, nums: list[int]) -> int:
        BIG = 10**5 + 1
        N = len(nums)
        for i in range(len(nums)):
            if nums[i] < 0 or nums[i] > N:
                nums[i] = 0

        for i in range(len(nums)):
            loc = (nums[i] % BIG) - 1
            if 0 <= loc < N:
                nums[loc] = (nums[loc] % BIG) + BIG

        for i in range(len(nums)):
            if nums[i] < BIG:
                return i + 1
        return N + 1
